### PR TITLE
Fix data leakage caused by reshuffling before splitting

### PR DESCRIPTION
**Pull request by aazuspan**
I thought those MSEs were too good to be true... @grovduck There's not much to review here, but just wanted to make you aware of the change and make sure this didn't raise any other red flags elsewhere in the code. I need to re-run things to get an idea of exactly how much the scores will drop and what that means for the different scale comparisons, but I don't think it's going to be pretty!

The Problem
-----------

To avoid artificial differences between our train/test/validation sets, we shuffle the entire dataset prior to splitting. When run, this creates a data loading pipeline that looks roughly like:

```
load data -> shuffle -> split
```

By default, the shuffle method will reshuffle on each iteration through the dataset, i.e. every epoch. This means that we get a different split every epoch, allowing samples to leak between sets, inflating model performance.

This issue is outlined here:

https://github.com/tensorflow/tensorflow/issues/59279

And pointed out in the comments to this SO answer that would introduce this leakage:

https://stackoverflow.com/a/51258695/17707273

The Solution
------------

Use `reshuffle_each_iteration=False` to prevent reshuffling on each iteration. The dataset will now be loaded, shuffled, and split only once.

### CODE REVIEW

**Review by grovduck — APPROVED** [review]
@aazuspan, I definitely understand the issue, but I'm curious why the default behavior would be to shuffle after each iteration.  The issue that you referenced made it seem as if the TF developers intended this (shuffle after epoch) behavior, but the initial OP only wanted a `Warning` raised. 

~Also, I'm wondering if there is value reshuffling the `train` and `validation` data, but keeping the `test` data completely separate.  You probably know better if that's a common pattern for TF modeling.~  I think I don't know what I'm talking about here.  It sounds like shuffling the training data is common practice, but keeping the validation data the same on each iteration is also the standard.  The validation data (rather than the testing data) is what the model fitting is evaluated on, correct?  I was thinking that was determined by the testing data, but it doesn't look like the notebooks are using `load_test` at this point. 

I'm also curious how you picked up on this?

**Comment by aazuspan** [comment]
> I'm curious why the default behavior would be to shuffle after each iteration

Great question, and I'm not really sure the answer. There are cases where reshuffling is preferable - training is theoretically more effective if the model sees the data in a different order every time - but to me it seems less intuitive and riskier. It looks like a lot of people are falling into the same trap, so hopefully they make some progress on that issue.

> The validation data (rather than the testing data) is what the model fitting is evaluated on, correct? I was thinking that was determined by the testing data, but it doesn't look like the notebooks are using load_test at this point.

Exactly. This seems to be the standard in deep learning, where you keep a holdout "test" set that is never used until the model is finalized, at which point you use it, rather than the validation set, to generate the final metrics. The idea is that even though the model never sees the validation set, you can effectively "overfit" to it by fine-tuning hyperparameters. The test set acts as a "real world" evaluation that's completely independent because neither the model nor the user has ever looked at it. I'm not sure how faithfully people stick to that approach in reality, but it's a nice idea at least!

> I'm also curious how you picked up on this?

I noticed the model was generating almost flawless height predictions in areas that were fully shaded, so I figured I'd either made a breakthrough in computer vision or accidentally leaked data - the latter seemed more likely! Luckily that issue popped up pretty quick when I looked up "tensorflow data leakage", because otherwise I'm not sure I ever would have tracked it down.

**Comment by grovduck** [comment]
> I figured I'd either made a breakthrough in computer vision

If anyone would have done it, it was you :smile:

**Comment by aazuspan** [comment]
Haha! Thanks for giving this a look!
